### PR TITLE
[Serve] Add Embedding Inference support with '/v1/embeddings' endpoint

### PR DESCRIPTION
**Pull request by xthomaswang**
## Summary
Add end to end embedding inference support with /v1/embeddings endpoint, also provide related test files.

### Key Changes:
 - Add `Qwen3EmbeddingModel` subclass with quantize functions to fix embedding model compilation 
 - Add `AsyncEmbeddingEngine` with auto model-type detection, batch prefill optimization, and chunked prefill fallback
 - Add `/v1/embeddings` endpoint similar to openai, has dimension truncation and base64 encoding support
 - Add `bert-bge` model type 
 - Add comprehensive tests (logit match, engine-level, server test)
 
### 1. Model compilation
- `qwen3_model.py`: add new `Qwen3EmbeddingModel` subclass that returns hidden states instead of logits via `prefill_to_last_hidden_states` and batch variants
-  `qwen3_quantization.py`: 4 new `*_embedding` quantize functions（`no_quant_embedding`, `group_quant_embedding`, `ft_quant_embedding`, `block_scale_quant_embedding`) that instantiate `Qwen3EmbeddingModel` 
- `model.py`: register `bert-bge`
- `bert_loader.py`: `hf_prefix` support for BGE weight name mapping

### 2. Inference engine
- `engine.py`: New `AsyncEmbeddingEngine`
  - Auto-detects encoder and decoder from compiled module functions
  - Encoder path: CLS/SEP token injection, truncation, cls/mean/last pooling
  - Decoder path: Last-token pooling with EOS token append(matches HF tokenizer behavior), batch prefill when total tokens fit in prefill_chunk_size, chunked prefill fallback for long sequences
  - L2 normalization, async support via ThreadPoolExecutor
- `engine_utils.py`: `extract_embedding_metadata()`, `load_embedding_params()`, `detect_embedding_model_type()`

### 3. API endpoint
- `openai_api_protocol.py`: `EmbeddingRequest` and `EmbeddingResponse` protocol classes
- `openai_entrypoints.py`: `POST /v1/embeddings` with dimension truncation and base64 encoding format
- `server_context.py`: Embedding engine lifecycle management
- `cli/serve.py` + `interface/serve.py`: `--embedding-model` / `--embedding-model-lib` CLI arguments

### 1. EOS Token Append for Decoder Models

MLC tokenizer does not add special tokens by default, but HF `sentence-transformers` does. Without appending EOS, last-token pooling extracts the wrong hidden state. After fix, cross-backend cosine similarity > 0.99.

**Before (without EOS):**

| | MLC-LLM | HuggingFace | Ollama |
|---|---|---|---|
| **MLC-LLM** | 1.0000 | 0.6663 | 0.6644 |
| **HuggingFace** | 0.6663 | 1.0000 | 0.9996 |
| **Ollama** | 0.6644 | 0.9996 | 1.0000 |

**After (with EOS):**

| | MLC-LLM | HuggingFace | Ollama |
|---|---|---|---|
| **MLC-LLM** | 1.0000 | 0.9996 | 0.9995 |
| **HuggingFace** | 0.9996 | 1.0000 | 0.9996 |
| **Ollama** | 0.9995 | 0.9996 | 1.0000 |

### 2. Batch Prefill Optimization

Multiple sequences share one KV cache via `kv_state_add_sequence` + `batch_prefill_to_last_hidden_states`. Single forward pass when total tokens ≤ `prefill_chunk_size`. 

### 3. Model Type Auto-Detection

- `embed` + `prefill_to_last_hidden_states` → **decoder**
- `prefill` only → **encoder**

No user configuration required.

### 4. Encoder Truncation

Inputs exceeding `prefill_chunk_size` (default 512) are truncated. Future improvement: sliding window + mean pooling strategy for better long-text support.

## Test Plan

**Model-level**: MLC vs HuggingFace logit match (cosine > 0.99 across 9 multilingual test texts)
```bash
MLC_QWEN3_EMB_HF_DIR=<hf_model> \
MLC_QWEN3_EMB_MODEL_DIR=<compiled_dir> \
MLC_QWEN3_EMB_MODEL_LIB=<model.dylib> \
  pytest tests/python/model/test_qwen3_embedding.py -v
```

**Engine-level**: 15 tests (shape, L2 norm, batch, similarity ranking, determinism, async, edge cases)
```bash
MLC_SERVE_EMBEDDING_MODEL_LIB=<model.dylib> \
  pytest -m engine tests/python/serve/test_embedding_engine.py -v
```

**Endpoint e2e**: HTTP endpoint tests (single/batch input, dimension truncation, base64, model fallback)
```bash
MLC_SERVE_EMBEDDING_MODEL_LIB=<model.dylib> \
  pytest -m endpoint tests/python/serve/server/test_embedding_server.py -v
```

**Cross-backend benchmark** (attached `benchmark.py`): Compares MLC vs HuggingFace vs Ollama on retrieval (EN/ZH/multilingual/code), STS, classification, MRL dimension truncation, and throughput. [benchmark.py](https://github.com/user-attachments/files/25502865/benchmark.py)

> **Note on quantization**: MLC-LLM uses q0f16 (fp16, ~1.1 GB), HuggingFace uses BFloat16 (~1.1 GB) — both unquantized. Ollama uses Q8_0 8-bit quantization (639 MB). Despite quantization differences, cross-backend cosine > 0.99.
```bash
# Start MLC server in terminal 1
python start_mlc_server.py
# Run benchmark in terminal 2
python benchmark.py  #auto-detect 3 backends
python benchmark.py --backends mlc hf       # without Ollama
python benchmark.py --skip-speed            # skip speed test
```

Results (Qwen3-Embedding-0.6B, Apple M1):

| | MLC-LLM | HuggingFace | Ollama |
|---|---|---|---|
| EN Retrieval (nDCG@10) | 1.0000 | 1.0000 | 1.0000 |
| ZH Retrieval (nDCG@10) | 1.0000 | 1.0000 | 1.0000 |
| Multilingual Retrieval (nDCG@10) | 1.0000 | 1.0000 | 1.0000 |
| Code Retrieval (nDCG@10) | 1.0000 | 1.0000 | 1.0000 |
| STS (Spearman rho) | 0.7796 | 0.7796 | 0.7796 |
| Classification | 100% | 100% | 100% |
| MRL dim=64 (nDCG@4) | 1.0000 | 1.0000 | 1.0000 |
| Short batch (32 texts/sec) | 110.9 | 114.8 | 44.5 |
| Long batch (8 texts/sec) | 3.1 | 3.8 | 7.0 |
| Single latency (ms) | 287.4 | 30.8 | 69.7 |
| Cross-backend cosine | - | 0.9996 | 0.9995 |

> Mini-benchmark on small datasets — for official scores, run full MTEB.

### CODE REVIEW

**Comment by xthomaswang** [comment]
@MasterJH5574 This PR adds end-to-end embedding inference support for mlc_llm. I am thinking mlc_llm could be a good engine for future agent framework, so embedding support could be important in this part. 

Would appreciate your review when you have a chance. Happy to walk through any part of the implementation.

**Comment by MasterJH5574** [comment]
@xthomaswang Thanks for the great work! I'll find time to review the PR. At the meantime, could you please check the existing reviews from gemini?

File: python/mlc_llm/interface/serve.py [hunk 1]
```diff
@@ -51,6 +52,8 @@ def serve(
     allow_origins: Any,
     allow_methods: Any,
     allow_headers: Any,
+    embedding_model: Optional[str] = None,
+    embedding_model_lib: Optional[str] = None,
```
Comment (MasterJH5574) [inline python/mlc_llm/interface/serve.py]: Could we follow other parameters and remove the default values?  And regarding the positions, I think it's better to put these two right below `additional_models`.

File: python/mlc_llm/cli/serve.py [hunk 2]
```diff
@@ -240,4 +252,6 @@ def main(argv):
         allow_origins=parsed.allow_origins,
         allow_methods=parsed.allow_methods,
         allow_headers=parsed.allow_headers,
+        embedding_model=parsed.embedding_model,
+        embedding_model_lib=parsed.embedding_model_lib,
```
Comment (MasterJH5574) [inline python/mlc_llm/cli/serve.py]: Same comment regarding the position.

File: python/mlc_llm/cli/serve.py [hunk 3]
```diff
@@ -198,6 +198,18 @@ def main(argv):
...
+        type=str,
+        default=None,
+        help="Path to the embedding model weight directory (enables /v1/embeddings endpoint)",
+    )
+    parser.add_argument(
+        "--embedding-model-lib",
+        type=str,
+        default=None,
+        help="Path to the compiled embedding model library (.so/.dylib file)",
+    )
```
Comment (MasterJH5574) [inline python/mlc_llm/cli/serve.py]: Same regarding the positions.

File: python/mlc_llm/serve/engine.py [hunk 4]
```diff
@@ -1965,3 +1972,386 @@ def _request_stream_callback_impl(
...
+        # Reuse existing utility: device detection
+        self.device = detect_device(device) if isinstance(device, str) else device
+        # Reuse existing utility: tokenizer
+        self.tokenizer = Tokenizer(model)
+
+        # Load TVM module, metadata, and params via engine_utils helpers
+        ex = tvm.runtime.load_module(model_lib)
+        vm = relax.VirtualMachine(ex, device=self.device)
+        self._mod = vm.module
+        self._metadata = engine_utils.extract_embedding_metadata(ex)
```
Comment (MasterJH5574) [inline python/mlc_llm/serve/engine.py]: We can get the metadata from `self._mod["_metadata"]` right? In that case we don't need `extract_embedding_metadata` to create a vm on CPU again to fetch the metadata.

File: python/mlc_llm/serve/engine_utils.py [hunk 5]
```diff
@@ -260,3 +260,81 @@ def __exit__(self, exc_type, exc_value, traceback) -> None:
...
+    """
+    from tvm.contrib import tvmjs  # pylint: disable=import-outside-toplevel
+
+    params, meta = tvmjs.load_tensor_cache(model_weight_path, device)
+    param_names = [param["name"] for param in model_metadata["params"]]
+    assert len(param_names) == meta["ParamSize"]
+    return [params[name] for name in param_names]
+
+
+def detect_embedding_model_type(mod) -> str:
```
Comment (MasterJH5574) [inline python/mlc_llm/serve/engine_utils.py]: Consider using `Literal["encoder", "decoder"]` given they are the only supported types.

File: python/mlc_llm/serve/engine.py [hunk 6]
```diff
@@ -1965,3 +1972,386 @@ def _request_stream_callback_impl(
...
+                pooled = output_np[0].mean(axis=0)
+            else:  # "last"
+                pooled = output_np[0, -1, :]
+
+            # L2 normalize
+            norm = np.linalg.norm(pooled)
+            if norm > 1e-12:
+                pooled = pooled / norm
+
+            embeddings.append(pooled.tolist())
```
Comment (MasterJH5574) [inline python/mlc_llm/serve/engine.py]: I'm wondering how much time this part usually takes. We can possibly use PyTorch to calculate given tvm-ffi provides fast conversion between tvm Tensor and PyTorch Tensor.

Same for other methods below.

File: tests/python/model/test_qwen3_embedding.py [hunk 7]
```diff
@@ -103,10 +103,13 @@ def _mlc_logits(text, tokenizer, mlc_module, params, metadata, dev):
...
     begin_forward(kv_cache, ShapeTuple([0]), ShapeTuple([seq_len]))
-    logits, _ = prefill_func(embedding, kv_cache, params)
+    hidden_states, _ = prefill_func(embedding, kv_cache, params)
     end_forward(kv_cache)
 
-    return logits.numpy()[0, -1, :]
+    # Compute logits from hidden states using embed_tokens weight (tie_word_embeddings)
+    hidden = hidden_states.numpy().astype(np.float32)
+    logits = hidden @ embed_weight.float().numpy().T
+    return logits[0, -1, :]
```
Comment (MasterJH5574) [inline tests/python/model/test_qwen3_embedding.py:112]: Same here. If we can leverage PyTorch I think it may be better.

File: python/mlc_llm/serve/engine.py [hunk 8]
```diff
@@ -1965,3 +1972,386 @@ def _request_stream_callback_impl(
...
+
+        When total tokens fit within prefill_chunk_size, all inputs are processed
+        in a single batch forward pass using shared KV cache. Otherwise, falls back
+        to sequential chunked prefill per input.
+        """
+        # Read KV cache config from metadata
+        sliding_window = self._metadata.get("sliding_window_size", -1)
+        context_window = self._metadata.get("context_window_size", 32768)
+        prefill_chunk = self._metadata.get("prefill_chunk_size", 2048)
+        max_seq_len = sliding_window if context_window == -1 else context_window
```
Comment (MasterJH5574) [inline python/mlc_llm/serve/engine.py]: We may an assertion check that `max_seq_len` is not -1, since it is used below.

File: python/mlc_llm/serve/engine.py [hunk 9]
```diff
@@ -1965,3 +1972,386 @@ def _request_stream_callback_impl(
...
+            ShapeTuple([prefill_chunk]),
+            ShapeTuple([16]),
+            ShapeTuple([support_sliding]),
+        )
+
+        # Register all sequences
+        seq_ids = list(range(batch_size))
+        seq_lens = [len(t) for t in token_lists]
+        for sid in seq_ids:
+            self._kv_state_add_sequence(kv_cache, sid)
```
Comment (MasterJH5574) [inline python/mlc_llm/serve/engine.py]: Do we need corresponding kv_state_remove_sequence? It looks to me that right now there is no removal.

File: python/mlc_llm/interface/serve.py [hunk 10]
```diff
@@ -85,6 +88,20 @@ def serve(
...
+        if embedding_model is not None:
+            if embedding_model_lib is None:
+                raise ValueError(
+                    "--embedding-model-lib is required when --embedding-model is specified."
+                )
+            emb_engine = AsyncEmbeddingEngine(
+                model=embedding_model,
+                model_lib=embedding_model_lib,
+                device=device,
+            )
```
Comment (MasterJH5574) [inline python/mlc_llm/interface/serve.py]: Nit: could we move these piece outside the ServerContext?

File: python/mlc_llm/serve/engine.py [hunk 11]
```diff
@@ -1965,3 +1972,386 @@ def _request_stream_callback_impl(
             batch_outputs.append(outputs)
             self.state.record_event(request_id, event="finish callback")
         return (batch_outputs, None)
+
+
+# ====== Embedding Engine ======
+
+
+class AsyncEmbeddingEngine:  # pylint: disable=too-many-instance-attributes
```
Comment (MasterJH5574) [inline python/mlc_llm/serve/engine.py]: Since this file is already very long, it might be better to have `AsyncEmbeddingEngine` in a standalone file `embedding_engine.py`.

File: python/mlc_llm/serve/engine.py [hunk 12]
```diff
@@ -1965,3 +1972,386 @@ def _request_stream_callback_impl(
...
+
+    def _init_decoder(self, model: str) -> None:
+        """Initialize decoder (Qwen3-Embeddings style) model functions."""
+        # Read EOS token from config — needed for last-token pooling to match HF behavior
+        self._decoder_eos_token_id: Optional[int] = None
+        config_path = os.path.join(model, "mlc-chat-config.json")
+        if os.path.exists(config_path):
+            with open(config_path, encoding="utf-8") as f:
+                chat_config = json.load(f)
+            self._decoder_eos_token_id = chat_config.get("eos_token_id")
```
Comment (MasterJH5574) [inline python/mlc_llm/serve/engine.py]: `eos_token_id` might be a list, so we may want to add some type checking or error reporting here.

https://github.com/mlc-ai/mlc-llm/blob/5c60c39c1fe30e6728914dbe23149c4fe0b624cd/python/mlc_llm/protocol/mlc_chat_config.py#L61

File: python/mlc_llm/interface/serve.py (hunk 1 above)
Comment (xthomaswang) [reply python/mlc_llm/interface/serve.py]: fixed

File: python/mlc_llm/cli/serve.py (hunk 2 above)
Comment (xthomaswang) [reply python/mlc_llm/cli/serve.py]: fixed

File: python/mlc_llm/cli/serve.py (hunk 3 above)
Comment (xthomaswang) [reply python/mlc_llm/cli/serve.py]: fixed

File: python/mlc_llm/serve/engine.py (hunk 4 above)
Comment (xthomaswang) [reply python/mlc_llm/serve/engine.py]: fixed

File: python/mlc_llm/serve/engine_utils.py (hunk 5 above)
Comment (xthomaswang) [reply python/mlc_llm/serve/engine_utils.py]: fixed

File: python/mlc_llm/serve/engine.py (hunk 6 above)
Comment (xthomaswang) [reply python/mlc_llm/serve/engine.py]: That's a great suggestion and totally makes sense intuitively!
 
I tested on MacBook M1 Pro. The post-processing (pooling + L2 normalize) takes ~0.25ms with NumPy, which is a tiny portion of total `embed()` time. I also compared PyTorch (`torch.from_dlpack`) with NumPy — PyTorch is actually slower (~0.55ms) on these small tensors.      
                                                                                                                                           
More importantly, the embedding engine calls Python → TVM VM directly (not through the C++ engine). `torch.from_dlpack()` is zero-copy so it holds a reference to the VM workspace buffer, which gets reused/overwritten on the next forward pass `.numpy()` copies to CPU and avoids this aliasing issue.

In the future, adding route embedding inferecen through the C++ engine like chat/completion could resolve the issue.

File: tests/python/model/test_qwen3_embedding.py (hunk 7 above)
Comment (xthomaswang) [reply tests/python/model/test_qwen3_embedding.py:112]: same as above

File: python/mlc_llm/serve/engine.py (hunk 8 above)
Comment (xthomaswang) [reply python/mlc_llm/serve/engine.py]: fixed

File: python/mlc_llm/serve/engine.py (hunk 9 above)
Comment (xthomaswang) [reply python/mlc_llm/serve/engine.py]: fixed

File: python/mlc_llm/interface/serve.py (hunk 10 above)
Comment (xthomaswang) [reply python/mlc_llm/interface/serve.py]: fixed

File: python/mlc_llm/serve/engine.py (hunk 11 above)
Comment (xthomaswang) [reply python/mlc_llm/serve/engine.py]: fixed

File: python/mlc_llm/serve/engine.py (hunk 12 above)
Comment (xthomaswang) [reply python/mlc_llm/serve/engine.py]: Done. Added type checking for eos_token_id (handles both int and List[int]) and expanded the EOS appending logic.

The EOS handling has 3 parts:
1. Tokenizer auto-append check: Read the `post_processor` from `tokenizer.json `to detect if the tokenizer automatically appends the EOS token natively.
2. Fallback: If the tokenizer doesn't handle it → read eos_token_id from config. If it's a list, take `eos[0]`, if it's an int, use directly.
3. Conditional append at inference:
  if (
      not `self._decoder_tokenizer_appends_eos`
      and `self._decoder_eos_token_id is not None`
      and `tokens[-1] != self._decoder_eos_token_id`
  )

File: python/mlc_llm/serve/engine.py (hunk 6 above)
Comment (MasterJH5574) [reply python/mlc_llm/serve/engine.py]: I see, thanks for explaining that! Let's go with numpy then.

**Comment by MasterJH5574** [comment]
@xthomaswang Thanks for addressing the comments. Could you also take a look at those from gemini? Feel free to resolve them if they don't make sense.

Update: never mind. I took a look and they seem to be out of scope.
